api/envoy/extensions/filters/http/proto_api_scrubber/v3/config.proto

```diff
@@ -0,0 +1,323 @@
+// }
+//
+// The filtered response output by this filter will be the following:
+//
+// .. code-block:: json
+//
+// {
+//   "book": {
+//     "title": "Book Title",
+//     "author": "Book Author",
+//     "publisher": "Book Publisher"
+//   }
+// }
+//
+// Note that the fields `debug_info` and `book.debug_info` are filtered out from
+// the response since the request entitlement of type "USER_TYPE" is PROD while
+// the restrictions on the fields `debug_info` and `book.debug_info` is DEV.
+
+// [#next-free-field: 6]
+message ProtoApiScrubberConfig {
+  // Specifies the filtering mode of this filter.
+  // Currently only ``OVERRIDE`` mode is supported.
+  enum FilteringMode {
+    // If unspecified, falls back to ``OVERRIDE``.
+    FILTERING_MODE_UNSPECIFIED = 0;
+
+    // Override the original request/response body with the filtered
+    // request/response body.
+    OVERRIDE = 1;
+  }
+
+  // The proto descriptor set for the proto services.
+  oneof descriptor_set {
+    // It could be passed by a local file through ``Datasource.filename`` or
+    // embedded in the ``Datasource.inline_bytes``.
+    config.core.v3.DataSource data_source = 1;
+
+    // Unimplemented, the key of proto descriptor TypedMetadata.
+    // Among filters depending on the proto descriptor, we can have a
+    // TypedMetadata for proto descriptors, so that these filters can share one
+    // copy of proto descriptor in memory.
+    string proto_descriptor_typed_metadata = 2;
+  }
+
+  // Data sources for the request entitlements.
+  // Entitlements are matched with restrictions of the same name at runtime.
+  // Make sure that the entitlement source is present for each of the
+  // corresponding restrictions.
+  // Key - entitlement name
+  // Value - entitlement source
+  map<string, RequestEntitlementSource> request_entitlement_sources = 3;
+
+  // Contains the restrictions for the API, methods, messages, fields, etc.
+  Restrictions restrictions = 4;
+
+  FilteringMode filtering_mode = 5;
+}
+
+// Specifies the source of entitlements for a request.
+// The value for each entitlement is expected to be a comma separated list of
+// strings.
+message RequestEntitlementSource {
+  enum Source {
+    // If unspecified, the entitlement will be empty string.
+    SOURCE_UNSPECIFIED = 0;
+
+    // The entitlement will be fetched from the request header ``key``.
+    REQUEST_HEADER = 1;
+
+    // The entitlement will be fetched from the filter state ``key``.
+    FILTER_STATE = 2;
+  }
+
+  Source source = 1;
+
+  // The key to lookup the entitlement from the specified ``source``.
+  string key = 2;
+}
+
+// Contains the restrictions for the API, methods, messages, fields, etc.
+message Restrictions {
+  // Specifies the method restrictions.
+  // Key - Fully qualified method name.
+  //     - ``${package}.${Service}.${Method}``, like
+  //     - ``endpoints.examples.bookstore.BookStore.GetShelf``
+  // Value - Method restrictions.
```

Comment:
its not clear how this will render - but almost certainly not how you want/expect